File: docker/backend/app/routers/qa.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel
from typing import Optional

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.dao import conversation_dao, message_dao
from app.services import qa_service, model_service

router = APIRouter()


def ok(data=None, message="ok"):
    return {"code": 200, "data": data, "message": message}


class AskRequest(BaseModel):
    conversation_id: Optional[int] = None
    question: str
    model_name: Optional[str] = None


class ConvRequest(BaseModel):
    title: str = "New Conversation"


@router.post("/ask")
async def ask_question(req: AskRequest, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    conv_id = req.conversation_id
    if not conv_id:
        conv = await conversation_dao.create(db, user.id, req.question[:50])
        await db.commit()
        conv_id = conv.id

    conv = await conversation_dao.get_by_id(db, conv_id)
    if not conv or conv.user_id != user.id:
        raise HTTPException(status_code=403, detail="无权访问该会话")

    # 自动更新对话标题（如果还是默认标题）
    if conv.title == "New Conversation":
        new_title = req.question[:50]
        await conversation_dao.update_title(db, conv_id, new_title)
        await db.commit()

    # Resolve model_name: use provided, or default from DB
    model_name = req.model_name
    if not model_name:
        from app.dao import model_dao
        default_config = await model_dao.get_default_config(db, "chat")
        model_name = default_config.model_name if default_config else "deepseek-chat"

    result = await qa_service.rag_pipeline(db, conv_id, req.question, model_name, user.id)
    result["conversation_id"] = conv_id
    return ok(result)


@router.get("/conversations")
async def list_conversations(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    convs = await conversation_dao.get_by_user(db, user.id)
    return ok([{"id": c.id, "title": c.title, "created_at": str(c.created_at)} for c in convs])


@router.post("/conversations")
async def create_conversation(req: ConvRequest = None, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    title = req.title if req else "New Conversation"
    conv = await conversation_dao.create(db, user.id, title)
    await db.commit()
    return ok({"id": conv.id, "title": conv.title})


@router.delete("/conversations/{conv_id}")
async def delete_conversation(conv_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    conv = await conversation_dao.get_by_id(db, conv_id)
    if not conv or conv.user_id != user.id:
        raise HTTPException(status_code=403, detail="无权访问该会话")
    await conversation_dao.delete_conv(db, conv_id)
    await db.commit()
    return ok(message="会话删除成功")


@router.get("/conversations/{conv_id}/messages")
async def get_messages(conv_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    conv = await conversation_dao.get_by_id(db, conv_id)
    if not conv or conv.user_id != user.id:
        raise HTTPException(status_code=403, detail="无权访问该会话")
    messages = await message_dao.get_by_conversation(db, conv_id)
    return ok([
        {
            "id": m.id, "role": m.role, "content": m.content,
            "sources": m.sources, "kg_references": m.kg_references,
            "model_name": m.model_name,
            "tokens_used": m.tokens_used, "created_at": str(m.created_at),
        }
        for m in messages
    ])


@router.get("/models")
async def get_models(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    models = await model_service.get_enabled_chat_models(db)
    return ok(models)

File: docker/frontend/src/stores/qa.js
/**
 * 智能问答状态管理 (Pinia Store)
 * 功能：管理问答会话状态，包括会话列表、当前会话、消息列表
 */

import { defineStore } from 'pinia'
import { ref } from 'vue'
import { getConversations, createConversation, deleteConversation, getMessages, askQuestion } from '../api/qa'

export const useQAStore = defineStore('qa', () => {
  // ========== 状态 ==========
  const conversations = ref([])    // 会话列表
  const currentConv = ref(null)    // 当前选中的会话
  const messages = ref([])         // 当前会话的消息列表
  const sending = ref(false)       // 是否正在发送消息

  // ========== 方法 ==========

  /**
   * 加载会话列表
   */
  async function loadConversations() {
    const res = await getConversations()
    conversations.value = res.data
  }

  /**
   * 创建新会话
   * @param {string} title - 会话标题（默认 'New Conversation'）
   */
  async function newConversation(title) {
    const res = await createConversation({ title: title || 'New Conversation' })
    const conv = res.data
    await loadConversations()
    await selectConversation(conv)
    return conv
  }

  /**
   * 删除会话
   * @param {number} id - 会话ID
   */
  async function removeConversation(id) {
    await deleteConversation(id)
    // 如果删除的是当前会话，清空消息列表
    if (currentConv.value?.id === id) {
      currentConv.value = null
      messages.value = []
    }
    await loadConversations()
  }

  /**
   * 选择会话并加载消息
   * @param {object} conv - 会话对象
   */
  async function selectConversation(conv) {
    currentConv.value = conv
    const res = await getMessages(conv.id)
    messages.value = res.data
  }

  /**
   * 发送消息
   * @param {string} question - 用户问题
   * @param {string} modelName - 使用的模型名称
   * @returns {object} - 包含 answer, sources, kg_references 等
   */
  async function sendMessage(question, modelName) {
    sending.value = true
    try {
      const res = await askQuestion({
        conversation_id: currentConv.value?.id,
        question,
        model_name: modelName || undefined,
      })
      // 如果是新会话，更新当前会话并刷新会话列表
      if (!currentConv.value) {
        currentConv.value = { id: res.data.conversation_id }
        await loadConversations()
      }
      // 重新加载消息列表（包含新发送的消息和AI回答）
      const msgRes = await getMessages(res.data.conversation_id)
      messages.value = msgRes.data
      return res.data
    } finally {
      sending.value = false
    }
  }

  return { conversations, currentConv, messages, sending, loadConversations, newConversation, removeConversation, selectConversation, sendMessage }
})
